// coolkit-api/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const store_1 = require("./store");
const api_1 = require("./api");
const utils_1 = require("./utils");
function init(params) {
    const { appId, appSecret, debug, useTestEnv = false, at, rt, countryCode, timeout } = params;
    store_1.setAppId(appId);
    store_1.setAppSecret(appSecret);
    store_1.setUseTestEnv(useTestEnv);
    if (debug) {
        store_1.setDebug(debug);
    }
    if (at) {
        store_1.setAt(at);
    }
    if (rt) {
        store_1.setRt(rt);
    }
    if (countryCode) {
        store_1.setDomain(utils_1.getDomainByCountryCode(countryCode));
    }
    if (timeout) {
        store_1.setTimeout(timeout);
    }
}
exports.default = {
    init,
    showStore: store_1.showStore,
    user: api_1.user,
    home: api_1.home,
    device: api_1.device,
    family: api_1.family,
    message: api_1.message,
    scene: api_1.scene,
    other: api_1.other,
    openPlatform: api_1.openPlatform,
    getCmsContent: utils_1.getCmsContent
};
