// coolkit-api/dist/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.openPlatform = exports.other = exports.scene = exports.message = exports.family = exports.device = exports.home = exports.user = void 0;
var user_1 = require("./user");
Object.defineProperty(exports, "user", { enumerable: true, get: function () { return user_1.user; } });
var home_1 = require("./home");
Object.defineProperty(exports, "home", { enumerable: true, get: function () { return home_1.home; } });
var device_1 = require("./device");
Object.defineProperty(exports, "device", { enumerable: true, get: function () { return device_1.device; } });
var family_1 = require("./family");
Object.defineProperty(exports, "family", { enumerable: true, get: function () { return family_1.family; } });
var message_1 = require("./message");
Object.defineProperty(exports, "message", { enumerable: true, get: function () { return message_1.message; } });
var scene_1 = require("./scene");
Object.defineProperty(exports, "scene", { enumerable: true, get: function () { return scene_1.scene; } });
var other_1 = require("./other");
Object.defineProperty(exports, "other", { enumerable: true, get: function () { return other_1.other; } });
var open_platform_1 = require("./open-platform");
Object.defineProperty(exports, "openPlatform", { enumerable: true, get: function () { return open_platform_1.openPlatform; } });
