// coolkit-api/dist/api/scene/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scene = void 0;
const store_1 = require("../../store");
const utils_1 = require("../../utils");
exports.scene = {
    getSceneList(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/list', 'POST', params, store_1.getAt());
        });
    },
    addScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2', 'POST', params, store_1.getAt());
        });
    },
    updateScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2', 'PUT', params, store_1.getAt());
        });
    },
    getSpecScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/thelist', 'POST', params, store_1.getAt());
        });
    },
    deleteScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest(`/v2/smartscene2?id=${params.id}`, 'DELETE', {}, store_1.getAt());
        });
    },
    sortScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/sort', 'POST', params, store_1.getAt());
        });
    },
    execScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/execute', 'POST', params, store_1.getAt());
        });
    },
    getExecHistory(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/history', 'POST', params, store_1.getAt());
        });
    },
    extSceneList(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/ext-scene', 'POST', params, store_1.getAt());
        });
    },
    getWebhookUrl() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/webhooks/url', 'GET', null, store_1.getAt());
        });
    },
    setWebhookAndScene(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/webhooks', 'POST', params, store_1.getAt());
        });
    },
    getWebhookAndScene() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/webhooks', 'GET', null, store_1.getAt());
        });
    },
    executeWebhook(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield utils_1.sendRequest('/v2/smartscene2/webhooks/execute', 'GET', params, store_1.getAt());
        });
    }
};
