// coolkit-ws/dist/class/WebSocket.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const getWsIpServices_1 = __importDefault(require("../services/getWsIpServices"));
const tools_1 = require("../utils/tools");
const isomorphic_ws_1 = __importDefault(require("isomorphic-ws"));
const eventEmitter_1 = __importDefault(require("./../utils/eventEmitter"));
const enum_1 = require("../ts/enum");
class WebSocketService {
    constructor(config) {
        WebSocketService.connectConfig = config;
    }
    _initWs() {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                const { apikey, region, useTestEnv = false } = WebSocketService.connectConfig;
                const { domain, port } = yield getWsIpServices_1.default(region, useTestEnv);
                const ws = new isomorphic_ws_1.default(`wss://${domain}:${port}/api/ws`);
                eventEmitter_1.default.once(`${enum_1.EEventType.INIT_WS}${apikey}`, (ev) => {
                    console.log('我收到握手成功的信息了');
                    resolve(ev);
                });
                ws.onclose = (ev) => this.onClose(ev);
                ws.onopen = (ev) => this.onOpen(ev);
                ws.onerror = (ev) => this.onError(ev);
                ws.onmessage = (ev) => this.onMessage(ev);
                WebSocketService.ws = ws;
            }));
        });
    }
    reconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            WebSocketService.ws = null;
            WebSocketService.isOpen = false;
            clearInterval(WebSocketService.hbInterval);
            WebSocketService.hbInterval = null;
            for (let count = 0; count < 10; count++) {
                yield this._initWs();
                if (WebSocketService.isOpen) {
                    break;
                }
            }
            if (!WebSocketService.isOpen) {
                throw new Error('长连接出现错误，重试失败，请重新连接！');
            }
        });
    }
    onOpen(ev) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('open==================');
            yield WebSocketService._sendHandShakeMsg();
        });
    }
    onError(ev) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('长连接出错，尝试重连', ev);
            yield this.reconnect();
        });
    }
    onClose(ev) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('长连接已关闭');
            if (WebSocketService.initClose) {
                WebSocketService.ws = null;
                WebSocketService.isOpen = false;
                clearInterval(WebSocketService.hbInterval);
                WebSocketService.hbInterval = null;
                return;
            }
            yield this.reconnect();
        });
    }
    onMessage(ev) {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = ev;
            if (data === 'pong') {
                console.log('ws 接收到心跳信息', data);
                return;
            }
            const decodedData = JSON.parse(data);
            if (decodedData.error === 0 && decodedData.config && decodedData.config.hb) {
                console.log('握手成功了');
                WebSocketService.isOpen = true;
                WebSocketService.sendMessage('ping');
                WebSocketService.hbInterval = setInterval(() => {
                    WebSocketService.sendMessage('ping');
                }, (decodedData.config.hbInterval - 7) * 1000);
                eventEmitter_1.default.emit(`${enum_1.EEventType.INIT_WS}${decodedData.apikey}`, {
                    error: 0,
                    msg: 'success',
                });
                return;
            }
            if (decodedData.error !== 0 && decodedData.actionName === 'userOnline') {
                console.log('长连接握手出错');
                return;
            }
            if ((decodedData.hasOwnProperty('error') && decodedData.deviceid) || decodedData.params) {
                const { sequence, deviceid } = decodedData;
                eventEmitter_1.default.emit(`${enum_1.EEventType.DEVICE_MSG}${sequence}${deviceid}`, decodedData);
            }
        });
    }
    static sendMessage(params) {
        if (WebSocketService.ws) {
            if (typeof params === 'string') {
                WebSocketService.ws.send(params);
                return;
            }
            WebSocketService.ws.send(JSON.stringify(params));
        }
    }
    static _sendHandShakeMsg() {
        return __awaiter(this, void 0, void 0, function* () {
            if (WebSocketService.ws) {
                let params;
                if (WebSocketService.connectConfig.userAgent === 'app') {
                    const { appid, userAgent, at, apikey } = WebSocketService.connectConfig;
                    params = {
                        action: 'userOnline',
                        version: 8,
                        ts: tools_1.createAppTs(),
                        at,
                        userAgent,
                        apikey,
                        appid,
                        nonce: tools_1.createNonce(),
                        sequence: `${Date.now()}`,
                    };
                }
                else {
                    const { userAgent, apikey, chipid, deviceid } = WebSocketService.connectConfig;
                    params = {
                        action: 'register',
                        version: 8,
                        ts: tools_1.createDeviceTs(),
                        userAgent,
                        apikey,
                        deviceid,
                        d_seq: `${Date.now()}`,
                    };
                    chipid && (params.chipid = chipid);
                }
                console.log(`SL : WebSocketService : _sendHandShakeMsg : params`, params);
                WebSocketService.sendMessage(params);
            }
        });
    }
    sendThing(deviceConfig, action) {
        return __awaiter(this, void 0, void 0, function* () {
            const { deviceid, params, ownerApikey } = deviceConfig;
            const sequence = `${Date.now()}`;
            const isApp = WebSocketService.connectConfig.userAgent === 'app';
            let deviceParams = {
                action,
                deviceid,
                apikey: ownerApikey,
                userAgent: isApp ? 'app' : 'device',
                sequence,
                params,
            };
            action === 'query' && isApp && (deviceParams.selfApikey = WebSocketService.connectConfig.apikey);
            console.log(`SL : Device : updateThing : deviceParams`, JSON.stringify(deviceParams));
            WebSocketService.sendMessage(deviceParams);
            return deviceParams;
        });
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            if (WebSocketService.ws) {
                return WebSocketService.ws;
            }
            return yield this._initWs();
        });
    }
    close() {
        if (WebSocketService.ws) {
            WebSocketService.initClose = true;
            WebSocketService.ws.close();
        }
    }
}
exports.default = WebSocketService;
WebSocketService.initClose = false;
