// coolkit-api-device/dist/store/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTimeout = exports.getTimeout = exports.getUseTestEev = exports.setUseTestEev = exports.getDebug = exports.setDebug = exports.getRt = exports.setRt = exports.getAt = exports.setAt = exports.getDomain = exports.setDomain = exports.getAppSecret = exports.setAppSecret = exports.getIsDevice = exports.setIsDevice = exports.getDeviceId = exports.setDeviceId = exports.setApiKey = exports.getApiKey = exports.showStore = void 0;
const store = {
    deviceId: '',
    appSecret: '',
    apiKey: '',
    domain: '',
    at: '',
    rt: '',
    debug: false,
    isDevice: false,
    useTestEnv: false,
    timeout: 0,
};
function showStore() {
    console.log(store);
}
exports.showStore = showStore;
function getApiKey() {
    return store.apiKey;
}
exports.getApiKey = getApiKey;
function setApiKey(v) {
    store.apiKey = v;
}
exports.setApiKey = setApiKey;
function setDeviceId(v) {
    store.deviceId = v;
}
exports.setDeviceId = setDeviceId;
function getDeviceId() {
    return store.deviceId;
}
exports.getDeviceId = getDeviceId;
function setIsDevice(v) {
    store.isDevice = v;
}
exports.setIsDevice = setIsDevice;
function getIsDevice() {
    return store.isDevice;
}
exports.getIsDevice = getIsDevice;
function setAppSecret(v) {
    store.appSecret = v;
}
exports.setAppSecret = setAppSecret;
function getAppSecret() {
    return store.appSecret;
}
exports.getAppSecret = getAppSecret;
function setDomain(v) {
    store.domain = v;
}
exports.setDomain = setDomain;
function getDomain() {
    return store.domain;
}
exports.getDomain = getDomain;
function setAt(v) {
    store.at = v;
}
exports.setAt = setAt;
function getAt() {
    return store.at;
}
exports.getAt = getAt;
function setRt(v) {
    store.rt = v;
}
exports.setRt = setRt;
function getRt() {
    return store.rt;
}
exports.getRt = getRt;
function setDebug(v) {
    store.debug = v;
}
exports.setDebug = setDebug;
function getDebug() {
    return store.debug;
}
exports.getDebug = getDebug;
function setUseTestEev(v) {
    store.useTestEnv = v;
}
exports.setUseTestEev = setUseTestEev;
function getUseTestEev() {
    return store.useTestEnv;
}
exports.getUseTestEev = getUseTestEev;
function getTimeout() {
    return store.timeout;
}
exports.getTimeout = getTimeout;
function setTimeout(v) {
    store.timeout = v;
}
exports.setTimeout = setTimeout;
